### ISSUE TITLE
Metamorphic pathfinder

### ISSUE THREAD

**Issue by aubergine10** [krzychu124/Cities-Skylines-Traffic-Manager-President-Edition#114]
Before I describe this in detail, which will take some time, are you guys crazy enough (you'd have to be!) to attempt writing a new path finder from scratch if the concept is good enough?

**_Very_ brief summary**

> Note: I'm describing this just as roads, but  it would also be ped paths, ferry lines, etc.

* Summarise the detailed information (lanes, restrictions, etc) at the segment level
* Summarise the line of segments between two junctions in to a single "road"
* At each junction, summarise the lane connections in to routes between each road at junction
* Result is a 'zoomed out' view of the transport network - basically a bunch of junctions with 'roads' connecting them.
* Pathfinder chooses route from A to B at the macroscopic level
* Once a route is chosen, the detailed path is worked out for that route

There's a huge bunch of stuff that the current pathfinder wades through at the microscopic (lane) level that it just doesn't need to do. The metamorphic pathfinder finds a route at the macroscopic level, then fills in the details for that route at the microscopic level.

The metamorphic pathfinder also has the advantage that most of the mods' features can be "always on" because they won't impact performance any more.

**Comment by Strdate** [comment]
This is a good idea, I had a similar one as well. But still, you probably don't need all vehicles to have the paths calculated in this way, only a portion.
Of course there are some catches when coding this, for example there can by way from segment A to B and from B to C, but there is no way from A to C because the vehicle cannot change lanes.
Before creating a new pathfinder we should see how the emergency vehicles behavior and all this will affect it.

**Comment by aubergine10** [comment]
Yes, I considered the 'lack off lane change' issue that will arise from the approach taken in OP. Thankfully, you've provided a very concise clarification of the problem. I don't know, yet, how to fully handle that issue. And there are similar issues, for example dealing with congestion. How do we encode these things in to the macroscopic summary of the transport network?

Ultimately, it's a math problem. It's almost like we'd need some way to encode information pertaining to the upcoming junction(s) in to the routing at the previous junction. At lest with the lane change issue, it's  predominantly a one-time calculation when networks (roads, etc) are updated. But how to deal with the congestion issue which changes due to a multitude of reasons (disasters, football matches, concerts, 'rush hour' if using RealTime mod, etc).

Those two issues, lane routing (taking in to account lane changes) and congestion, are the things that cause my brain to melt then pour out of my ears and scamper across the floor in a bid to escape to the nearest mountains in the hopes of peacefully herding goats for the rest of its life so as not to be traumatised by such ghoulish maths.

**Comment by aubergine10** [comment]
Oh, as for emergency vehicles, they route normally but they can do fancy stuff on the way. Although not as fancy as these garbage trucks: https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/64#issuecomment-468088631

**Comment by FireController1847** [comment]
As ambitious of a project this is, I'd be in for it if you and @krzychu124 are. This is probably super duper long-term, though, and won't probably be gotten to until like a year from now due to all the changes we're making in terms of the UI and other logic (ex. emergency vehicle)

**Comment by VictorPhilipp** [comment]
I think your suggested method is called "Contraction hierarchies" (https://en.wikipedia.org/wiki/Contraction_hierarchies). I have seen an implementation of it in an OSM routing framework called "GraphHopper" (https://github.com/graphhopper/graphhopper).
It requires heavy precomputations and the graphhopper implementation is quite inflexible, e.g. no dynamic costs are supported. However I have not looked further into it.

**Comment by RenaKunisaki** [comment]
Is the current pathfinder GPU accelerated? Could this one be?

**Comment by matfax** [comment]
I think that would give a more realistic approach if done correctly. I'm not sure but I think there is an issue or a Steam discussion on the topic that the AI always chooses the "perfect" route or shortest path and not the route you would normally choose in real life. You would prefer highways and ring roads just because they have fewer "obstacles", they are more comfortable to drive, have fewer junctions and traffic lights, you rather follow the street signs.

This concept could be projected to a multilayered graph. 
```
Outer Zone { Street { Segment { Lane } } }
Districts { District { Street { Segment { Lane } } } }
```

I have not mentioned junctions because I see them just as segments that also belong to streets and have multiple lanes with a source and a sink and prohibited lane change.

Assuming we need to reach a certain building, it would be sufficient to know the nearest lane to the building, numbers are not necessary. From there you can then look for a parking lot.

So, for instance, this could be our target:
```
Residental District -> Beach Avenue -> Segment 15 -> Lane 2
```

Concerning the pathfinding, I see two scenarios. The obvious one is to predetermine the complete or partial macroscopic path as a subgraph for every car. That has the advantage that lane changing is more realistic given that subgraph and cars can gauge the lane changes considering traffic density and where they are heading.

But there are many cars and considering #316 you could approach it from another perspective. Assuming that there are fewer junctions than cars, it would be more memory efficient to let the junctions know "their street signs", or where the cars need to go to get to the desired destination. Of course, "normal" connections between two segments should not count as junctions, where lane numbers match and are not modified via TM:PE UI.

When a car starts driving, it asks the next "real" junction upfront giving its desired destination object. Technically, it won't even ask, it just adds itself to the next segment's lane, and the segment knows the next junctions and does all the work. Getting the request, the junction then looks into its KV-Store or cache and either returns the matched target lanes or it lazily uses the pathfinder to find the appropriate lanes if there is no match. The target lanes identify the lanes before the junction to get to the required lanes on the junction itself that lead to the target. The car then stores the target lanes and eventually switch the lane on its way to the junction. Once the car has reached the junction via the lanes, the junction directs it via one of the possible junction lanes that lead to the target. After that, the car repeats the process and asks the next junction.

Junctions would have a small caches for every layer and also know their own macroscopic position.
```
Position: Workers District -> Busy Street
Districts:
  Outer Zone -> target lane 1, 2; junction lane 1, 2, 3
  Residental District -> target lane 2; junction lane 3
Streets of this District: ...
Segments of this Street: ...
```
It will return target lane 2 for the example above.

The next junction is inside the Residental District, so it will now skip to the street lookup.
```
Position: Residental District -> Town Street
Districts: ...
Streets of this District: 
  Beach Avenue -> target lane 1, junction lane 1, 2
  Lonely Hills -> target lane 1, junction lane 3
Segments of this Street: ...
```
Return: target lane 1

What do you think about it thus far?

**Comment by kvakvs** [comment]
This article on pathfinding https://arxiv.org/ftp/arxiv/papers/1810/1810.01776.pdf
mentions some ways to optimize searches by defining key points and some dividing lines between the regions of the map. And there's a theorem that the solution would cross the divider between the two closest points: from A to the dividing line, and from B to the dividing line.

**Comment by aubergine10** [comment]
Not read the pdf yet, but I'd pondered something like that should we get a node/segment grid up and running. Draw straight line between source and target, then half way draw perpendicular line across it. Look for routes crossing the perpendicular line and do some sort of quick path find (as pondered in earlier comments above) to find if any of those routes can reach target, then check if source can reach those routes. Using the node/segment grid we could create a rectangle to define the width of the area to examine, and if no routes are found just move that rectangle left or right (from perspective of source facing target) and try again, possibly caching that "going form here to there needs to take a detour to the left" if possible.
